Add guide for LLM eval (#10058)
mlflow/mlflow: docs/source/llms/llm-evaluate/index.rst

            ],
        }
    )

    with mlflow.start_run() as run:
        system_prompt = "Answer the following question in two sentences"
        # Wrap "gpt-4" as an MLflow model.
        logged_model_info = mlflow.openai.log_model(
            model="gpt-4",
            task=openai.ChatCompletion,
            artifact_path="model",
            messages=[
                {"role": "system", "content": system_prompt},
                {"role": "user", "content": "{question}"},
            ],
        )

        # Use predefined question-answering metrics to evaluate our model.
        results = mlflow.evaluate(
            logged_model_info.model_uri,
            eval_data,
            targets="ground_truth",
            model_type="question-answering",
        )
        print(f"See aggregated evaluation results below: \n{results.metrics}")

        # Evaluation result for each data record is available in `results.tables`.
        eval_table = results.tables["eval_results_table"]
        print(f"See evaluation table below: \n{eval_table}")

LLM Evaluation Metrics
=======================

There are two types of LLM evaluation metrics in MLflow:

> 1. Metrics relying on SaaS model (e.g., OpenAI) with prompt engineering, e.g., :py:func:`mlflow.metrics.relevance`. These  

Reviewer: can we say that this model is for scoring purposes.

   one prompt consisting of the following information to the SaaS model, and extract the score from model response:

   * Metrics definition.
   * Metrics grading criteria.
   * Reference examples.
   * Input data/context.
   * Model output.
   * [optional] Ground truth.

   More details of how these fields are set can be found in the section "Create your Custom LLM-evaluation Metrics".

2. Function-based per-row metrics. These metrics calculate a score for each data record (row in terms of Pandas/Spark dataframe),
   based on certain functions, like Rouge (:py:func:`mlflow.metrics.rougeL`) or Flesch Kincaid (:py:func:`mlflow.metrics.flesch_kincaid_grade_level`). 
   These metrics are similar to traditional metrics.

Select Metrics to Evaluate
--------------------------

MLflow LLM evaluation includes default collections of metrics for pre-selected tasks, e.g, "question-answering". Depending on the 
type of LLM use case that you are evaluating, these pre-defined collections can greatly simplify the process of running evaluations. 

The supported metrics for given model types are shown below:

> * **question-answering**:

Reviewer: token count is included for all 3 of these types. can we add that

Author: hmm that's not evaluation metrics right? I feel it's just some extra information computed along with evaluating.

Reviewer: while it may be a bit different semantically, its computed as an evaluation metric so i think we should include it here.

An example of using the predefined metrics for a given ``model_type`` is shown below:

.. code-block:: python

    results = mlflow.evaluate(
        model,
        eval_data,
        targets="ground_truth",
        model_type="question-answering",
    )

1. Specify a custom list of metrics by explicitly referencing a metric calculation function.

    * To add additional metrics to the default collection from part 1 above, add the function names to the ``extra_metrics`` argument.
    * To diable default metric calculation and only calculate explicit metrics, remove the ``model_type`` argument and define the desired metrics. 

An example of disabling the default metrics and explicitly declaring a subset of metrics to calculate is shown below:

.. code-block:: python

    results = mlflow.evaluate(
        model,
        eval_data,
        targets="ground_truth",
        extra_metrics=[mlflow.metrics.toxicity(), mlflow.metrics.latency()],
    )

The full reference for supported evaluation metrics can be found `here <../python_api/mlflow.html#mlflow.evaluate>`_. 

> 

Reviewer: @chenmoneygithub we need to add a section here where we discuss the pre-canned LLM judged metrics such as faithfulness and answer_similarity. can we add that here?

Author: sure!

Create your Custom LLM-evaluation Metrics
---------------------------------------------

> Create LLM-as-judge Evaluation Metrics (Cateogory 1)

Reviewer: ```suggestion
Create LLM-as-judge Evaluation Metrics (Category 1)
```

            "some informal professional settings."
            "- Score 2: Language is overall formal but still have casual words/phrases. Borderline for professional contexts."
            "- Score 3: Language is balanced and avoids extreme informality or formality. Suitable for most professional contexts. "
            "- Score 4: Language is noticeably formal, respectful, and avoids casual elements. Appropriate for formal "
            "business or academic settings. "
        ),
        examples=[professionalism_example_score_2, professionalism_example_score_4],
        model="openai:/gpt-3.5-turbo-16k",
        parameters={"temperature": 0.0},
        aggregations=["mean", "variance"],
        greater_is_better=True,
    )

Best Practices for Creating Custom Metrics (LLM as the Judge)
""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""

Under the hood of evaluting with LLM as the judge is prompt engineering, and how to find the best prompt is still under research. 
Here are some tips for setting required fields of custom SaaS LLM evaluation metrics:

* TODO(sunish): please share your learnings here.

> * TODO(prithvi): please share your learnings here.

Reviewer: Lets add this later. This will take some time to produce.

* TODO(ann): please share your learnings here.

> Create Per-row LLM Evluation Metrics (Cateogory 2)

Reviewer: ```suggestion
Create Per-row LLM Evaluation Metrics (Category 2)
```

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

This is very similar to creating a custom traditional metrics, with the exception of returning a `EvaluationResult` instance.
Basically you need to:

1. Implement a ``eval_fn`` to define your scoring logic, it must take in 3 args ``predictions``, ``targets`` and ``metrics``.
   ``eval_fn`` must return a :py:func:`mlflow.metrics.MetricValue` instance.

> 2. Pass ``eval_fn`` and other arguments to ``mlflow.metricsmake_metric`` API to create the metric. 

Reviewer: ```suggestion
2. Pass ``eval_fn`` and other arguments to ``mlflow.metrics.make_metric`` API to create the metric. 
```

            evaluators="default",
        )
        print(f"See aggregated evaluation results below: \n{results.metrics}")

        eval_table = results.tables["eval_results_table"]
        print(f"See evaluation table below: \n{eval_table}")

View Evaluation Results
========================

View Evaluation Results via Code
-----------------------------------

``mlflow.evaluate()`` returns the evaluation results as an :py:func:`mlflow.models.EvaluationResult` instace. 
To see the score on selected metrics, you can check:

* ``metrics``: stores the aggregated results, like average/variance across the evaluation dataset. Let's take a second
  pass on the code example above and focus on printing out the aggregated results.
  
  .. code-block:: python

    with mlflow.start_run() as run:
        results = mlflow.evaluate(
            data=eval_data,
            targets="ground_truth",
            predictions="predictions",
            extra_metrics=[mlflow.metrics.answer_similarity()],
            evaluators="default",
        )
        print(f"See aggregated evaluation results below: \n{results.metrics}")
* ``tables``: stores the per-row evaluation results. 

>   TODO(prithvi): The code example always uses "eval_results_table" as the key to fetch eval table, is this a hard requirement?

Reviewer: yes. the table produced by `mlflow.evaluate` will be called `eval_results_table`

MLflow provides an API :py:func:`mlflow.evaluate()` to help evaluate your LLMs.

MLflow's LLM evaluation functionality consists of 3 main components:

1. **A model to evaluate**: it can be an MLflow ``pyfunc`` model, a URI pointing to one registered 
   MLflow model, or any python callable that represents your model, e.g, a HuggingFace text summarization pipeline. 
2. **Metrics**: the metrics to compute, LLM evaluate will use LLM metrics. 
3. **Evaluation data**: the data your model is evaluated at, it can be a pandas Dataframe, a python list, a 
   numpy array or an :py:func:`mlflow.data.dataset.Dataset` instance.

Quickstart
==========

Below is a simple example that gives an quick overview of how MLflow LLM evaluation works. The example builds
a simple question-answering model by wrapping "openai/gpt-4" with custom prompt. You can paste it to
your IPython or local editor and execute it, and install missing dependencies as prompted. Running the code 
requires OpenAI API key, if you don't have an OpenAI key, you can set it up [here](https://platform.openai.com/account/api-keys).

.. code-block:: shell

>     export OPENAI_API_KEY = [YOUR_OPEN_AI_KEY]

Reviewer 2: ```suggestion
    export OPENAI_API_KEY=YOUR_OPEN_AI_KEY
```

            "business or academic settings. "
        ),
        examples=[professionalism_example_score_2, professionalism_example_score_4],
        model="openai:/gpt-3.5-turbo-16k",
        parameters={"temperature": 0.0},
        aggregations=["mean", "variance"],
        greater_is_better=True,
    )

..
    TODO(prithvi): add best practice for creating GenAI metrics.

Create Per-row LLM Evluation Metrics (Category 2)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

This is very similar to creating a custom traditional metrics, with the exception of returning a `EvaluationResult` instance.
Basically you need to:

1. Implement a ``eval_fn`` to define your scoring logic, it must take in 3 args ``predictions``, ``targets`` and ``metrics``.
   ``eval_fn`` must return a :py:func:`mlflow.metrics.MetricValue` instance.
2. Pass ``eval_fn`` and other arguments to ``mlflow.metricsmake_metric`` API to create the metric. 

The following code creates a dummy per-row metric called ``"over_10_chars"``: if the model output is greater than 10, 
the score is 1 otherwise 0.

.. code-block:: python

    def eval_fn(predictions, targets, metrics):

>         scores = []
>         for i in range(len(predictions)):
>             if len(predictions[i]) > 10:
>                 scores.append(1)
>             else:
>                 scores.append(0)
>         return MetricValue(
>             scores=scores,
>             aggregate_results=standard_aggregations(scores),
>         )

Reviewer 2: ```suggestion
        scores = []
        for pred in predictions:
            scores.append(len(pred) > 10)
        return MetricValue(
            scores=scores,
            aggregate_results=standard_aggregations(scores),
        )
```

Author: hmm I think that gives True/False, although it is compatible with int, I don't want to leave users an impression that they can use boolean as the return value.
